### ISSUE TITLE
Unclear documentation of `as_cvmat` can cause bugs

### ISSUE THREAD

**Issue by AntoniosBarotsis** [OmkarKabadagi5823/cv-bridge-rs#1]
I recently came across an interesting issue in a project of mine that makes use of this crate. I initially thought it might be an issue with the underlying opencv bindings but after an [extensive discussion with the bindings' crate author](https://github.com/twistedfall/opencv-rust/issues/277#issuecomment-1548000923), he figured out that the issue lies in the unsafe block of the [`as_cvmat`](https://github.com/OmkarKabadagi5823/cv-bridge-rs/blob/56109baa971ff4adde0b5aa6a99c7bf5dd206937/src/cv_image.rs#L188C12-L209) function.

> The `cv-bridge-rs` crate uses the unsafe function `new_rows_cols_with_data()` to create a `Mat` in `as_cvmat()`. This makes the resulting `Mat` a reference to the data contained in `CvImage`. As soon as the `CvImage` is destroyed the memory is freed and `Mat` is now referring to that freed memory. [...]

I feel like having the option of using the current `as_cvmat` function is great for performance reasons, however, this quirk needs to be explicitly documented. Maybe it's even worth marking the function as `unsafe` if this is not an invariant you can guarantee (which would be a breaking change). I feel like just cloning is not really a great solution for the library itself, even though that is how I currently circumvent the issue in my code. I'm not sure if you can make use of some other opencv function (maybe [`from_exact_iter`](https://docs.rs/opencv/latest/opencv/prelude/struct.Mat.html#method.from_exact_iter) would work?).

I'm not entirely sure how this can be prevented in this crate's code as I have basically never written any unsafe Rust. If there is something I could help with, I am willing to do so considering how useful I've found his crate :)

**Comment by OmkarKabadagi5823** [comment]
Thank you for your interest in this project. I apologize for responding to this so late. Coming to the issue, I  noticed it before but did not care to fix it earlier. One way to solve this is to have two different functions. In one, we return `Mat` referencing the data in `CvImage`. The other could consume or copy the data in `CvImage` to `Mat`.  The issue of `Mat` possibly referring to a freed memory could be fixed using lifetimes. `from_exact_iter`  could be used in case copying is necessary. 

If you would like to contribute, please feel free to submit a pull request with your changes.

**Comment by AntoniosBarotsis** [comment]
I don't have too much free time for now but I could look into this in a little over a month perhaps. Just to reiterate what you mentioned to make sure we are on the same page:

- One function returns a `Mat` that references `CvImage` and ties the lifetimes of the two
- Another consumes the `Mat` and copies the data internally as a more expensive, non-lifetime alternative

**Comment by OmkarKabadagi5823** [comment]
Yes, you have summarized the two functions correctly. I think it would be useful to have both options available for different use cases. I could fix the issue with the `as_cvmat` function using lifetimes soon. I will look into `from_exact_iter` or any other function which could be used for consuming or copying data to `Mat`. If you have time to work on this in the future, that would be great. Please let me know if you have any questions or need any help.

**Comment by AntoniosBarotsis** [comment]
Yeah having both options makes sense. As I mentioned I could work on this from around mid-july, if you make any changes before that could you mention them here? :)

**Comment by OmkarKabadagi5823** [comment]
Sure.

**Comment by AntoniosBarotsis** [comment]
I just started looking into this and I was wondering if there is some easy way to test it. I tried making the following example to reproduce the bug:

```rs
use cv_bridge::CvImage;
use opencv::{prelude::Mat, imgcodecs::{imread, IMREAD_UNCHANGED}};

fn main() {
  let img = imread("./test.jpg", IMREAD_UNCHANGED).unwrap();
  let mat = CvImage::from_cvmat(img, "rgb8").unwrap().as_cvmat().unwrap();
  // show_in_window(&img);
  show_in_window(&mat);
}

fn show_in_window(img: &Mat) {
  opencv::highgui::imshow("img", &img).expect("open window");
  let _res = opencv::highgui::wait_key(0).expect("keep window open");
}
```

however, passing `mat` to that function results in garbage for me (even though commenting it out and using `img` instead works fine)

![image](https://github.com/OmkarKabadagi5823/cv-bridge-rs/assets/50240570/6a8df1f8-104f-45e6-b6b7-2388f095a82a)

Nothing bad should be happening yet though as far as I can tell? I'm probably doing something silly but can't quite figure out what. It would definitely be a good idea to have a reproducible example in place before we work on this.
